### PR TITLE
Type safe middleware

### PR DESCRIPTION
**Pull request by Andrew-Lees11**
Adds the new TypeSafeMiddleware protocol and routes for up to 3 TypeSafeMiddleware in Codable routing.

## Description
TypeSafeMiddleware is a protocol for middleware where they return an instance of self instead of altering the request and response. This allows them to be used in Codable routing.

The routes have been added in a new file for Codable routing allowing a user to have up to three type safe middleware on a route by declaring them in there route signature.

## Motivation and Context
type safe middleware means application gains compile-time safety and developers benefit from enhanced separation of concerns between the middleware(s) and route handlers. 

It also allows middleware such as Auth and sessions them to be used in Codable routes.

## How Has This Been Tested?
Has been tested manually using Sessions as a typed middleware.
Automated tests still need to be added.
Appropriate Docs also need to be added describing the new routes.

### CODE REVIEW

**Comment by shmuelk** [comment]
Why is one limited to three middleware? Why isn't the list of middleware an array?

File: Sources/Kitura/CodableRouter+TypedMiddleware.swift [hunk 1]
```diff
@@ -0,0 +1,1020 @@
+/*
```
Comment (ianpartridge) [inline Sources/Kitura/CodableRouter+TypedMiddleware.swift]: Can this file be called `CodableRouter+TypeSafeMiddleware.swift`?

File: Sources/Kitura/CodableRouter+TypedMiddleware.swift [hunk 2]
```diff
@@ -0,0 +1,1020 @@
...
+import KituraNet
+import KituraContracts
+import Dispatch
+
+// Codable router
+
+extension Router {
+
+    // Used by PUT and PATCH with identifier
+    public typealias MiddlewareIdentifierCodableClosure<T: TypeSafeMiddleware, Id: Identifier, I: Codable, O: Codable> = (T, Id, I, @escaping CodableResultClosure<O>) -> Void
```
Comment (ianpartridge) [inline Sources/Kitura/CodableRouter+TypedMiddleware.swift]: Is there a reason why these typealiases are inside `extension Router {}`? I would have thought they should be top-level.

File: Sources/Kitura/CodableRouter+TypedMiddleware.swift [hunk 3]
```diff
@@ -0,0 +1,1020 @@
...
+    
+    // POST using two middleware
+    fileprivate func postSafely<T1: TypeSafeMiddleware, T2: TypeSafeMiddleware, I: Codable, O: Codable>(_ route: String, handler: @escaping TwoMiddlewareCodableClosure<T1, T2, I, O>) {
+        post(route) { request, response, next in
+            Log.verbose("Received POST type-safe request")
+            guard let codableInput = CodableHelpers.readCodableOrSetResponseStatus(I.self, from: request, response: response) else {
+                next()
+                return
+            }
+            self.handleMiddleware(T1.self, T2.self, request: request, response: response) { TypeSafeMiddleware1, TypeSafeMiddleware2 in
```
Comment (ianpartridge) [inline Sources/Kitura/CodableRouter+TypedMiddleware.swift]: Please name variables with lower-case names, e.g. `typeSafeMiddleware1` not `TypeSafeMiddleware1` to distinguish them from type names.  This comment applies across the whole PR.

**Comment by ianpartridge** [comment]
@shmuelk I think the feature we really want is [variadic generics](https://github.com/apple/swift/blob/master/docs/GenericsManifesto.md#variadic-generics) but they aren't here yet unfortunately. Andrew and Dave did explore using an array, I'll let them comment on why they decided against that approach.

**Comment by Andrew-Lees11** [comment]
Type Aliases (if we want to keep them) have been moved to a kitura contracts [PR](https://github.com/IBM-Swift/KituraContracts/pull/23) This has failed though since contracts doesn't know what typeSafeMiddleware are and we don't want it to import Kitura.

File: Sources/Kitura/RouterMiddleware.swift [hunk 4]
```diff
@@ -14,6 +14,8 @@
  * limitations under the License.
  */
 
+import KituraContracts
```
Comment (ianpartridge) [inline Sources/Kitura/RouterMiddleware.swift]: I think we can revert all the changes to this file now?

File: Sources/Kitura/TypeSafeMiddleware.swift [hunk 5]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright IBM Corporation 2016
```
Comment (ianpartridge) [inline Sources/Kitura/TypeSafeMiddleware.swift]: 2018

File: Sources/Kitura/TypeSafeMiddleware.swift [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    ///                     HTTP request.
+    /// - Parameter response: The `RouterResponse` object used to respond to the
+    ///                     HTTP request.
+    /// - Parameter completion: The closure to invoke once middleware processing is
+    ///                         complete. Either an instance of Self or a RequestError
+    ///                         should be provided, indicating a successful or failed
+    ///                         attempt to process the request, respectively.
+    static func handle(request: RouterRequest, response: RouterResponse, completion: @escaping (Self?, RequestError?) -> Void) -> Void
+    
+    /// Decribe the type-safe middleware
```
Comment (ianpartridge) [inline Sources/Kitura/TypeSafeMiddleware.swift]: Describe not Decribe. Also, please give an example of the kind of description that would be appropriate.

File: Sources/Kitura/CodableRouter+TypeSafeMiddleware.swift [hunk 7]
```diff
@@ -0,0 +1,890 @@
+/*
+ * Copyright IBM Corporation 2017
```
Comment (ianpartridge) [inline Sources/Kitura/CodableRouter+TypeSafeMiddleware.swift]: 2018

File: Sources/Kitura/CodableRouter+TypeSafeMiddleware.swift [hunk 8]
```diff
@@ -825,6 +844,7 @@ extension Router {
     ) {
         T.handle(request: request, response: response) { (typeSafeMiddleware: T?, error: RequestError?) in
             guard let typeSafeMiddleware = typeSafeMiddleware else {
+                print("failed typeSafeMiddleware. Error: \(String(describing: error))")
```
Comment (ianpartridge) [inline Sources/Kitura/CodableRouter+TypeSafeMiddleware.swift]: print!

File: Sources/Kitura/TypeSafeMiddleware.swift [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+     TODO: This seems to only apply to auth for the moment. Should this be in TypeSafeCredentials?
+     ### Usage Example: ###
+     The following is the description for basic auth from https://swagger.io/docs/specification/authentication/basic-authentication/:
+     ```
+     basicAuth:     # <-- arbitrary name for the security scheme
+        type: http
+        scheme: basic
+     ```
+     */
+    static func describe() -> String
```
Comment (ianpartridge) [inline Sources/Kitura/TypeSafeMiddleware.swift]: My only question is whether we should take this out for now.

File: Sources/Kitura/TypeSafeMiddleware.swift [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import KituraContracts
+
+/// Defines the protocol which all Kitura type-safe middleware must implement.
+///
+/// TypeSafeMiddleware are classes or structs that process an incoming request
+/// and on success create an instance of self, which is passed to the users route handler.
```
Comment (helenmasters) [inline Sources/Kitura/TypeSafeMiddleware.swift]: users route handler -> user's route handler

File: Sources/Kitura/CodableRouter+TypeSafeMiddleware.swift [hunk 11]
```diff
@@ -0,0 +1,1245 @@
...
+import Foundation
+import LoggerAPI
+import KituraNet
+import KituraContracts
+
+// Type-safe middleware Codable router
+
+extension Router {
+
+    // MARK: Codable Routing With Type-Safe Middleware
```
Comment (helenmasters) [inline Sources/Kitura/CodableRouter+TypeSafeMiddleware.swift]: At least the With should be lowercase in the // MARK here as it looks strange with a capital 'W' underneath the previous MARK'd title of Parameter handling (which has a lowercase 'h').

File: Sources/Kitura/CodableRouter+TypeSafeMiddleware.swift [hunk 12]
```diff
@@ -0,0 +1,1245 @@
...
+     ```
+     #### Multiple Middleware: ####
+     The closure can process up to three `TypeSafeMiddleware` objects by defining them in the handler:
+     ```swift
+     router.get("/user") { (middle1: Middle1, middle2: Middle2, middle3: Middle3, query: Query,
+                            respondWith: ([User]?, RequestError?) -> Void) in
+     ```
+     - Parameter route: A String specifying the URL path that will invoke the handler.
+     - Parameter handler: A closure that receives a TypeSafeMiddleware instance and a QueryParams instance, and returns an array of Codable objects or a RequestError.
+     */    public func get<T: TypeSafeMiddleware, Q: QueryParams, O: Codable>(
```
Comment (helenmasters) [inline Sources/Kitura/CodableRouter+TypeSafeMiddleware.swift]: This */ is being displayed in the API doc.  `public func...` needs to move onto the next line to fix this.

File: Sources/Kitura/CodableRouter+TypeSafeMiddleware.swift [hunk 13]
```diff
@@ -0,0 +1,1245 @@
...
+    }
+
+    /**
+     Sets up a closure that will be invoked when a POST request to the provided route is received by the server.
+     The closure accepts a successfully executed instance of `TypeSafeMiddleware`, a Codable object, an identifier
+     and a handler which responds with a single Codable object or a `RequestError`.
+     The handler contains the developer's logic, which determines the server's response.
+     ### Usage Example: ###
+     In this example, MySession is a struct that conforms to the TypeSafeMiddleware protocol.
+     session is an instance of MySession, that contains [Int: User] dictionary.
```
Comment (helenmasters) [inline Sources/Kitura/CodableRouter+TypeSafeMiddleware.swift]: You need to add the word 'an' before [Int: User] dictionary

File: Sources/Kitura/CodableRouter+TypeSafeMiddleware.swift [hunk 14]
```diff
@@ -0,0 +1,1245 @@
...
+     }
+     ```
+     #### Multiple Middleware: ####
+     The closure can process up to three `TypeSafeMiddleware` objects by defining them in the handler:
+     ```swift
+     router.post("/user") { (middle1: Middle1, middle2: Middle2, middle3: Middle3, user: User,
+                             id: Int, respondWith: (User?, RequestError?) -> Void) in
+     ```
+     - Parameter route: A String specifying the URL path that will invoke the handler.
+     - Parameter handler: A closure that receives a TypeSafeMiddleware instance a Codable object and an identifier, and returns a Codable object or a RequestError.
```
Comment (helenmasters) [inline Sources/Kitura/CodableRouter+TypeSafeMiddleware.swift]: Needs a comma after the word instance.

File: Sources/Kitura/CodableRouter+TypeSafeMiddleware.swift [hunk 15]
```diff
@@ -0,0 +1,1245 @@
...
+    // MARK: Codable Routing With Type-Safe Middleware
+
+    /**
+     Sets up a closure that will be invoked when a GET request to the provided route is received by the server.
+     The closure accepts a successfully executed instance of `TypeSafeMiddleware`
+     and a handler which responds with a single Codable object or a `RequestError`.
+     The handler contains the developer's logic, which determines the server's response.
+     ### Usage Example: ###
+     In this example, MySession is a struct that conforms to the TypeSafeMiddleware protocol.
+     session is an instance of MySession, that contains optional Users.
```
Comment (helenmasters) [inline Sources/Kitura/CodableRouter+TypeSafeMiddleware.swift]: Anywhere where we are starting a sentence with a variable name which is lowercase it looks poor in the jazzy API doc.

In this example, `MySession` is a struct that conforms to the `TypeSafeMiddleware` protocol and specifies an optional `User`, where `User` conforms to `Codable`.

**Comment by Andrew-Lees11** [comment]
Tests for all routes and documentation has now been added.
Two and Three middleware are marked as no doc so they don't show up in jazzy for clarity.

**Pull request merged**